routing/helicopter_router.cpp

```diff
@@ -0,0 +1,139 @@
+    +-------+--------+-----------------+---------------+
+    | S     | 1      | 0               | 1             |
+    | 1     | 1      | 1               | 2             |
+    | 1     | 2      | 2               | 3             |
+    | 2     | 2      | 3               | 4             |
+    | 2     | F      | 4               | 6             |
+    | 2     | F      | 4               | 6             |
+    +-------+--------+-----------------+---------------+
+
+  Constraints:
+  *  m_geometry.size() == 2 * checkpoints.size()
+  *  m_routeSegments.size() == m_geometry.size()-1
+  *  m_subroutes.size() == m_routeSegments.size()-1
+
+ */
+RouterResultCode HelicopterRouter::CalculateRoute(Checkpoints const & checkpoints,
+                                                      m2::PointD const & startDirection,
+                                                      bool adjustToPrevRoute,
+                                                      RouterDelegate const & delegate, Route & route)
+{
+  vector<m2::PointD> const & points = checkpoints.GetPoints();
+  geometry::Altitude const mockAltitude = 0;
+  vector<RouteSegment> routeSegments;
+  vector<double> times;
+  times.reserve(points.size()*2-1);
+  Segment const segment(kFakeNumMwmId, 0, 0, false);
+
+  for (size_t i = 0; i < points.size(); ++i)
+  {
+    turns::TurnItem turn(i, turns::PedestrianDirection::None);
+    geometry::PointWithAltitude const junction(points[i], mockAltitude);
+    RouteSegment::RoadNameInfo const roadNameInfo;
+
+    auto routeSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+    routeSegments.push_back(move(routeSegment));
+    times.push_back(0);
+
+    if (i == points.size() - 1)
+    {
+      // Create final segment.
+      turn = turns::TurnItem(i+1, turns::PedestrianDirection::ReachedYourDestination);
+      RouteSegment lastSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+      routeSegments.push_back(move(lastSegment));
+      times.push_back(0);
+    }
+    else if (i > 0)
+    {
+      // Duplicate intermidiate points.
+      RouteSegment intermidiateSegment = RouteSegment(segment, turn, junction, roadNameInfo);
+      routeSegments.push_back(move(intermidiateSegment));
+      times.push_back(0);
+    }
+  }
+
+  FillSegmentInfo(times, routeSegments);
+  route.SetRouteSegments(move(routeSegments));
+
+  vector<Route::SubrouteAttrs> subroutes;
+  for(size_t i = 1; i < points.size(); ++i)
+  {
+    if (i<points.size()-1)
+    {
+      auto subrt1 = Route::SubrouteAttrs(geometry::PointWithAltitude(points[i-1], mockAltitude),
+                                        geometry::PointWithAltitude(points[i], mockAltitude), i*2-2, i*2-1);
+      subroutes.push_back(move(subrt1));
+
+      auto subrt2 = Route::SubrouteAttrs(geometry::PointWithAltitude(points[i-1], mockAltitude),
+                                        geometry::PointWithAltitude(points[i], mockAltitude), i*2-1, i*2);
+      subroutes.push_back(move(subrt2));
```

Comment:
```
using PointWA = geometry::PointWithAltitude;

subroutes.emplace_back(PointWA(points[i-1], mockAltitude), PointWA(points[i-1], mockAltitude), i*2-2, i*2-1);

...
```

Reply:
Even better:
```
auto const ToPointWA = [](m2::PointD const & p)
{
  return geometry::PointWithAltitude(p, 0 /* altitude */);
};

subroutes.emplace_back(ToPointWA(points[i-1]), ToPointWA(points[i-1]), i*2-2, i*2-1);
```

Reply:
And what about performance? Would compiler be able to inline lambda here?

Reply:
Yes

Reply:
Fixed using your code